### PR TITLE
ES/03 checkpoint

### PR DESCRIPTION
**Pull request by samarjeet**
# ALCHEMI Toolkit Pull Request

Exact checkpoint / restore: Zarr checkpoint format; BaseDynamics state_dict / load_state_dict / redistribute_state / apply_thermodynamic_state for NVTLangevin and NVTNoseHoover; transactional manifest; round-trip test 

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

-
-
-

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/enhanced_sampling/_checkpoint.py
Comment (dallasfoster) [inline nvalchemi/enhanced_sampling/_checkpoint.py:1]: Writing state transactionally and refusing to restore a torn store has nothing
to do with enhanced sampling. Training already implements it, MD restart wants
it, and NEB restart will want it. Shipping a second implementation under
`enhanced_sampling/` is what guarantees a third.

```python
# nvalchemi/_checkpoint.py
class Stateful(Protocol):
    def state_dict(self) -> Mapping[str, Any]: ...
    def load_state_dict(self, state: Mapping[str, Any]) -> None: ...

def save_checkpoint(path, components: Mapping[str, Stateful], *,
                    batch: Batch | None = None, compatibility=None) -> None: ...

def load_checkpoint(path, components: Mapping[str, Stateful], *, device=None) -> None: ...
```

Hooks, integrators, biases and strategies all satisfy `Stateful` already, or
would under row 3. If refactoring `training/_checkpoint.py` onto this is too
much for the series, the minimum ask is that the new module lands at
`nvalchemi/_checkpoint.py` rather than inside `enhanced_sampling/`.

**Review by dallasfoster — CHANGES_REQUESTED** [review]
